### PR TITLE
Initial constant strain rate implementation

### LINKED ISSUE #44: Add constant strain rate mode
I had a user reach out about needing constant strain rate for some of their exploratory studies. It seems like I could probably follow something like what was discussed in the comment thread at https://github.com/LLNL/ExaConstit/pull/26#discussion_r559089039 .

It'll probably take a little bit of work to get a POC going, but I don't think it should be too awful with the changing BCs functionality now within the code.

As a first pass, the easiest version will be something like a forward eulerian version of things where we calculate the  velocity conditions based on the previous time step nodal locations. After we get that looking, we can look at more expensive versions like ones that do an iterative approach to calculate at the end time step or  mid-time step values.

Additionally, I plan on allowing this took work with meshes that don't  have a min at origin. Also, I plan on requiring the user to define what degree  of freedoms are free and which ones are constrained to the velocity gradient  condition.

As part of this set of calculations, I figured I would also calculate the Eulerian strain for the user as we can easily get out the deformation gradient, and so that value isn't hard to get at all. I figure this would make it much easier for them when calculating post-processing values as they'll automatically have the "true"  strain values. If they want the lagrangian strain tensor, they can just get that from the deformation gradient...

### PR DESCRIPTION
**Pull request by rcarson3**
This initial implementation adds a constant strain rate method based on what's outlined in #44 . This methods performs something akin to an explicit constant strain rate in that we use the initial time step node locations to calculate what the velocity should be for the defined constrained DOFs.  I would say that this is fine for an initial implementation. If we want something that is based on the end time step node locations or even the middle time step node locations that can be added later, but it will require additional work as we'll have to do a simple iterative solve for things.

After thinking about  this some more, I believe this will be a pretty powerful feature once added as it should allow for some cool simulations to be run like torsional tests.

I still need to add a few small things first before this can be merged or should be reviewed:

- [x] Need unit tests one for monotonic and one that mimics the cyclic example
- [x] Examine adding a new unit test that looks at having a torsional test case (shouldn't be too hard to implement with this capability)
- [x] Need to use a node offset in calculations that takes into account the minimum mesh location in order to allow for meshes that don't have their minimum coordinate the origin
- [x] Add some documentation on how to use the feature

### CODE REVIEW

**Comment by rcarson3** [comment]
As of https://github.com/LLNL/ExaConstit/pull/45/commits/c66f7093bb5b59a60f2eb6cd4bf99793742604c9 things work fairly well. However, I realized the torsional example really won't work with my current implementation. In order to get things to work, I will need to allow both constant strain rates and the previous boundary condition methods to exist at the same time.  I don't think this will be hard given the current abstraction, but it will introduce breaking changes to input decks...

Essentially, we would abuse the sign of the `BCs.essential_comps` to define if we are running constant strain rates or the old method for a given boundary. In order to do this, we would need to move the all constrain condition to be positive and set it to be 7 rather than -1.

**Comment by rcarson3** [comment]
Okay now with https://github.com/LLNL/ExaConstit/pull/45/commits/3879613c79f2aebf88779dccdb5a637d362acf4d it should be possible to have both applied velocity and applied velocity gradient conditions applied at the same time. The caveat of course is that for a given boundary it is one or the other you can't have both. I'll work on a torsion example/test next, but all the other tests appear to be now passing and I even added a new mixed mode test which shows the two work together.

**Comment by rcarson3** [comment]
Torsion example is now working (4 4th order elements), but I'll want to think about adding an option to all users to specify an origin / min set of points rather than having us calculate it for them.

torsion_example

**Comment by rcarson3** [comment]
Add an option to specify the "origin" / offset term used to calculate velocity gradient BCs calculations. By doing this, you can now do torsion like simulations as seen below. I'll add very small torsion test next now that I've shown this works to very large strains.
torsion_centered

**Comment by rcarson3** [comment]
@nrbertin this is the PR related to the constant strain rate additions I added for you. As this will introduce breaking changes to the options file, I was wondering if there was anything in the BCs side of things that would make your life easier? For example, I'm thinking of changing the `essential_vals` field to be a 2D array rather than the 1D array to make inputting things a bit easier for people?

You should be able to see the current changes I've made related to the input file in the `src/options.toml` file in the files changed tab of this PR.

**Comment by rcarson3** [comment]
Should pull this in after #42 -> #50 are merged in

File: test/data/voce_ea_cs.toml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+[BCs]
+    # Optional - tells the program that we'll have changing BCs
+    changing_ess_bcs = false
+    # Optional - tells the program if we want a constant strain rates rather than
+    # constant velocity boundary conditions
+    constant_strain_rate = true
+    #essential BC ids for the whole boundary
+    essential_ids = [1, 2, 3, 4]
+    #component combo (x,y,z = -1, x = 1, y = 2, z = 3, xy = 4, yz = 5, xz = 6, free = 0)
+    essential_comps = [-3, -1, -2, -3]
```
Comment (nrbertin) [inline test/data/voce_ea_cs.toml:53]: What are the negative values of essential_comps for here? Does it relate to the constant strain rate option?

**Review by nrbertin — APPROVED** [review]
The constant strain rate option is a great addition to the code

File: test/data/voce_ea_cs.toml (hunk 1 above)
Comment (rcarson3) [reply test/data/voce_ea_cs.toml:53]: Yes, the negatives here signify that the essential boundary condition should for a given boundary should have the velocity gradient / constant strain rate BC applied rather than the applied velocity BC.

Comment (rcarson3) [reply test/data/voce_ea_cs.toml:53]: Went through and updated all the test options files so this outdated info should no longer be in there

**Comment by rcarson3** [comment]
Thanks for testing it out @nrbertin and pushing me to get this working.
